File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/InvoiceDto.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.taktik.icure.entities.IdentityDocumentReader;
import org.taktik.icure.entities.Payment;
import org.taktik.icure.services.external.rest.v1.dto.embed.PaymentType;
import org.taktik.icure.services.external.rest.v1.dto.embed.InvoicingCodeDto;

public class InvoiceDto extends IcureDto {
	private Long invoiceDate; //yyyyMMdd
	private Long sentDate;
	private Long printedDate;

	private Double paid;
	private PaymentType paymentType;
	private List<Payment> payments;

	private List<InvoicingCodeDto> invoicingCodes;
	private String invoiceType;
	private String sentMediumType;
	private String interventionType;

	private String groupId;

	private String correctiveInvoiceId;
	private String correctedInvoiceId;

	private String recipientType; //org.taktik.icure.entities.HealthcareParty, org.taktik.icure.entities.Insurance, org.taktik.icure.entities.Patient
	private String recipientId; //for hcps and insurance, patient link happens through secretForeignKeys

	private String invoiceReference;
	private String thirdPartyReference;

	private String thirdPartyPaymentJustification;
	protected String thirdPartyPaymentReason;

	protected String gnotionNihii;
	protected String gnotionSsin;
	protected String gnotionLastName;
	protected String gnotionFirstName;
	protected String gnotionCdHcParty;

	protected Integer invoicePeriod;

	protected String internshipNihii;
	protected String internshipSsin;
	protected String internshipLastName;
	protected String internshipFirstName;
	protected String internshipCdHcParty;
	protected String internshipCbe;

	protected String supervisorNihii;
	protected String supervisorSsin;
	protected String supervisorLastName;
	protected String supervisorFirstName;
	protected String supervisorCdHcParty;
	protected String supervisorCbe;

	protected Integer longDelayJustification;

	protected Boolean creditNote;
	protected String creditNoteRelatedInvoiceId;

	protected String careProviderType;

	protected String error;

	protected String encounterLocationName;
	protected String encounterLocationNihii;
	protected Integer encounterLocationNorm;


	protected Map<String,String> receipts = new HashMap<>();

	protected IdentityDocumentReader idDocument;

	public Long getInvoiceDate() {
		return invoiceDate;
	}

	public void setInvoiceDate(Long invoiceDate) {
		this.invoiceDate = invoiceDate;
	}

	public Long getPrintedDate() {
		return printedDate;
	}

	public void setPrintedDate(Long printedDate) {
		this.printedDate = printedDate;
	}

	public List<InvoicingCodeDto> getInvoicingCodes() {
		return invoicingCodes;
	}

	public void setInvoicingCodes(List<InvoicingCodeDto> invoicingCodes) {
		this.invoicingCodes = invoicingCodes;
	}

	public String getRecipientType() {
		return recipientType;
	}

	public void setRecipientType(String recipientType) {
		this.recipientType = recipientType;
	}

	public String getRecipientId() {
		return recipientId;
	}

	public void setRecipientId(String recipientId) {
		this.recipientId = recipientId;
	}

	public String getInvoiceReference() {
		return invoiceReference;
	}

	public void setInvoiceReference(String invoiceReference) {
		this.invoiceReference = invoiceReference;
	}

	public String getCorrectiveInvoiceId() {
		return correctiveInvoiceId;
	}

	public void setCorrectiveInvoiceId(String correctiveInvoiceId) {
		this.correctiveInvoiceId = correctiveInvoiceId;
	}

	public String getCorrectedInvoiceId() {
		return correctedInvoiceId;
	}

	public void setCorrectedInvoiceId(String correctedInvoiceId) {
		this.correctedInvoiceId = correctedInvoiceId;
	}

	public String getInvoiceType() {
		return invoiceType;
	}

	public void setInvoiceType(String invoiceType) {
		this.invoiceType = invoiceType;
	}

	public String getSentMediumType() { return sentMediumType; }

	public void setSentMediumType(String sentMediumType) { this.sentMediumType = sentMediumType; }

	public String getInterventionType() { return interventionType; }

	public void setInterventionType(String interventionType) { this.interventionType = interventionType; }

	public String getGroupId() { return groupId; }

	public void setGroupId(String groupId) { this.groupId = groupId; }

	public Long getSentDate() {
		return sentDate;
	}

	public void setSentDate(Long sentDate) {
		this.sentDate = sentDate;
	}

	public String getThirdPartyReference() {
		return thirdPartyReference;
	}

	public void setThirdPartyReference(String thirdPartyReference) {
		this.thirdPartyReference = thirdPartyReference;
	}

	public Map<String, String> getReceipts() {
		return receipts;
	}

	public void setReceipts(Map<String, String> receipts) {
		this.receipts = receipts;
	}

	public Double getPaid() {
		return paid;
	}

	public void setPaid(Double paid) {
		this.paid = paid;
	}

	private String encryptedSelf;
	public String getEncryptedSelf() {
		return encryptedSelf;
	}

	public void setEncryptedSelf(String encryptedSelf) {
		this.encryptedSelf = encryptedSelf;
	}

	public String getGnotionNihii() { return gnotionNihii; }

	public void setGnotionNihii(String gnotionNihii) { this.gnotionNihii = gnotionNihii; }

	public String getGnotionSsin() { return gnotionSsin; }

	public void setGnotionSsin(String gnotionSsin) { this.gnotionSsin = gnotionSsin; }

	public String getGnotionLastName() { return gnotionLastName; }

	public void setGnotionLastName(String gnotionLastName) { this.gnotionLastName = gnotionLastName; }

	public String getGnotionFirstName() { return gnotionFirstName; }

	public void setGnotionFirstName(String gnotionFirstName) { this.gnotionFirstName = gnotionFirstName; }

	public String getGnotionCdHcParty() { return gnotionCdHcParty; }

	public void setGnotionCdHcParty(String gnotionCdHcParty) { this.gnotionCdHcParty = gnotionCdHcParty; }

	public Integer getInvoicePeriod() { return invoicePeriod;}

	public void setInvoicePeriod(Integer invoicePeriod) { this.invoicePeriod = invoicePeriod; }

	public String getInternshipNihii() { return internshipNihii; }

	public void setInternshipNihii(String internshipNihii) { this.internshipNihii = internshipNihii; }

	public String getInternshipSsin() { return internshipSsin; }

	public void setInternshipSsin(String internshipSsin) { this.internshipSsin = internshipSsin; }

	public String getInternshipLastName() { return internshipLastName; }

	public void setInternshipLastName(String internshipLastName) { this.internshipLastName = internshipLastName; }

	public String getInternshipFirstName() { return internshipFirstName; }

	public void setInternshipFirstName(String internshipFirstName) { this.internshipFirstName = internshipFirstName; }

	public String getInternshipCdHcParty() { return internshipCdHcParty; }

	public void setInternshipCdHcParty(String internshipCdHcParty) { this.internshipCdHcParty = internshipCdHcParty; }

	public String getInternshipCbe() { return internshipCbe; }

	public void setInternshipCbe(String internshipCbe) { this.internshipCbe = internshipCbe; }

	public String getSupervisorNihii() {
		return supervisorNihii;
	}

	public void setSupervisorNihii(String supervisorNihii) {
		this.supervisorNihii = supervisorNihii;
	}

	public String getSupervisorSsin() {
		return supervisorSsin;
	}

	public void setSupervisorSsin(String supervisorSsin) {
		this.supervisorSsin = supervisorSsin;
	}

	public String getSupervisorLastName() {
		return supervisorLastName;
	}

	public void setSupervisorLastName(String supervisorLastName) {
		this.supervisorLastName = supervisorLastName;
	}

	public String getSupervisorFirstName() {
		return supervisorFirstName;
	}

	public void setSupervisorFirstName(String supervisorFirstName) {
		this.supervisorFirstName = supervisorFirstName;
	}

	public String getSupervisorCdHcParty() {
		return supervisorCdHcParty;
	}

	public void setSupervisorCdHcParty(String supervisorCdHcParty) {
		this.supervisorCdHcParty = supervisorCdHcParty;
	}

	public void setSupervisorCbe(String supervisorCbe) { this.supervisorCbe = supervisorCbe; }

	public String getSupervisorCbe() { return supervisorCbe; }

	public Integer getLongDelayJustification() {
		return longDelayJustification;
	}

	public void setLongDelayJustification(Integer longDelayJustification) { this.longDelayJustification = longDelayJustification; }

	public Boolean getCreditNote() { return creditNote; }

	public void setCreditNote(Boolean creditNote) { this.creditNote = creditNote; }

	public String getCreditNoteRelatedInvoiceId() { return creditNoteRelatedInvoiceId; }

	public void setCreditNoteRelatedInvoiceId(String creditNoteRelatedInvoiceId) { this.creditNoteRelatedInvoiceId = creditNoteRelatedInvoiceId; }

	public String getCareProviderType() { return careProviderType; }

	public void setCareProviderType(String careProviderType) { this.careProviderType = careProviderType; }

	public String getError() { return error; }

	public void setError(String error) { this.error = error; }

	public String getEncounterLocationName() { return encounterLocationName; }

	public void setEncounterLocationName(String encounterLocationName) { this.encounterLocationName = encounterLocationName; }

	public String getEncounterLocationNihii() { return encounterLocationNihii; }

	public void setEncounterLocationNihii(String encounterLocationNihii) { this.encounterLocationNihii = encounterLocationNihii; }

	public Integer getEncounterLocationNorm() {
		return encounterLocationNorm;
	}

	public void setEncounterLocationNorm(Integer encounterLocationNorm) {
		this.encounterLocationNorm = encounterLocationNorm;
	}


	public String getThirdPartyPaymentJustification() {
		return thirdPartyPaymentJustification;
	}

	public void setThirdPartyPaymentJustification(String thirdPartyPaymentJustification) {
		this.thirdPartyPaymentJustification = thirdPartyPaymentJustification;
	}

	public String getThirdPartyPaymentReason() {
		return thirdPartyPaymentReason;
	}

	public void setThirdPartyPaymentReason(String thirdPartyPaymentReason) {
		this.thirdPartyPaymentReason = thirdPartyPaymentReason;
	}

	public PaymentType getPaymentType() {
		return paymentType;
	}

	public void setPaymentType(PaymentType paymentType) {
		this.paymentType = paymentType;
	}

	public List<Payment> getPayments() {
		return payments;
	}

	public void setPayments(List<Payment> payments) {
		this.payments = payments;
	}

	public IdentityDocumentReader getIdDocument() {
		return idDocument;
	}

	public void setIdDocument(IdentityDocumentReader idDocument) {
		this.idDocument = idDocument;
	}
	
}

File: src/main/kotlin/org/taktik/icure/services/external/rest/v1/dto/InvoiceDto.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.services.external.rest.v1.dto

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.github.pozo.KotlinBuilder
import io.swagger.v3.oas.annotations.media.Schema
import org.taktik.icure.services.external.rest.v1.dto.base.CodeStubDto
import org.taktik.icure.services.external.rest.v1.dto.base.EncryptableDto
import org.taktik.icure.services.external.rest.v1.dto.base.ICureDocumentDto
import org.taktik.icure.services.external.rest.v1.dto.base.StoredDocumentDto
import org.taktik.icure.services.external.rest.v1.dto.embed.*

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
@Schema(description = """This entity is a root level object. It represents an Invoice. It is serialized in JSON and saved in the underlying iCure CouchDB database.""")
data class InvoiceDto(
	@Schema(description = "The Id of the Invoice. We encourage using either a v4 UUID or a HL7 Id.") override val id: String,
	@Schema(description = "The revision of the invoice in the database, used for conflict management / optimistic locking.") override val rev: String? = null,
	override val created: Long? = null,
	override val modified: Long? = null,
	override val author: String? = null,
	override val responsible: String? = null,
	override val medicalLocationId: String? = null,
	override val tags: Set<CodeStubDto> = emptySet(),
	override val codes: Set<CodeStubDto> = emptySet(),
	override val endOfLife: Long? = null,
	override val deletionDate: Long? = null,

	@Schema(description = "The timestamp (unix epoch in ms) when the invoice was drafted, will be filled automatically if missing. Not enforced by the application server.") val invoiceDate: Long? = null, // yyyyMMdd
	@Schema(description = "The timestamp (unix epoch in ms) when the invoice was sent, will be filled automatically if missing. Not enforced by the application server.") val sentDate: Long? = null,
	@Schema(description = "The timestamp (unix epoch in ms) when the invoice is printed, will be filled automatically if missing. Not enforced by the application server.") val printedDate: Long? = null,
	val invoicingCodes: List<InvoicingCodeDto> = emptyList(),
	@Schema(description = "") val receipts: Map<String, String> = emptyMap(),
	@Schema(description = "The type of user that receives the invoice, a patient or a healthcare party") val recipientType: String? = null, // org.taktik.icure.services.external.rest.v1.dto.HealthcarePartyDto,

	// org.taktik.icure.services.external.rest.v1.dto.InsuranceDto, org.taktik.icure.services.external.rest.v1.dto.PatientDto
	@Schema(description = "Id of the recipient of the invoice. For healthcare party and insurance, patient link happens through secretForeignKeys") val recipientId: String? = null, // for hcps and insurance, patient link happens through secretForeignKeys
	val invoiceReference: String? = null,
	val thirdPartyReference: String? = null,
	val thirdPartyPaymentJustification: String? = null,
	val thirdPartyPaymentReason: String? = null,
	val reason: String? = null,
	@Schema(description = "The format the invoice should follow based on the recipient which could be a patient, mutual fund or paying agency such as the CPAS") val invoiceType: InvoiceTypeDto? = null,
	@Schema(description = "Medium of the invoice: CD ROM, Email, paper, etc.") val sentMediumType: MediumTypeDto? = null,
	val interventionType: InvoiceInterventionTypeDto? = null,
	val groupId: String? = null,
	@Schema(description = "Type of payment, ex: cash, wired, insurance, debit card, etc.") val paymentType: PaymentTypeDto? = null,
	val paid: Double? = null,
	val payments: List<PaymentDto>? = null,
	val gnotionNihii: String? = null,
	val gnotionSsin: String? = null,
	val gnotionLastName: String? = null,
	val gnotionFirstName: String? = null,
	val gnotionCdHcParty: String? = null,
	val invoicePeriod: Int? = null,
	val careProviderType: String? = null,
	val internshipNihii: String? = null,
	val internshipSsin: String? = null,
	val internshipLastName: String? = null,
	val internshipFirstName: String? = null,
	val internshipCdHcParty: String? = null,
	val internshipCbe: String? = null,
	val supervisorNihii: String? = null,
	val supervisorSsin: String? = null,
	val supervisorLastName: String? = null,
	val supervisorFirstName: String? = null,
	val supervisorCdHcParty: String? = null,
	val supervisorCbe: String? = null,
	val error: String? = null,
	val encounterLocationName: String? = null,
	val encounterLocationNihii: String? = null,
	val encounterLocationNorm: Int? = null,
	val longDelayJustification: Int? = null,
	val correctiveInvoiceId: String? = null,
	val correctedInvoiceId: String? = null,
	val creditNote: Boolean? = null,
	val creditNoteRelatedInvoiceId: String? = null,
	val idDocument: IdentityDocumentReaderDto? = null,

	//efact hospitalization
	val admissionDate: Long? = null,
	val locationNihii: String? = null,
	val locationService: Int? = null,

	//eattest cancel
	val cancelReason: String? = null,
	val cancelDate: Long? = null,
	val options: Map<String, String> = emptyMap(),

	override val secretForeignKeys: Set<String> = emptySet(),
	override val cryptedForeignKeys: Map<String, Set<DelegationDto>> = emptyMap(),
	override val delegations: Map<String, Set<DelegationDto>> = emptyMap(),
	override val encryptionKeys: Map<String, Set<DelegationDto>> = emptyMap(),
	override val encryptedSelf: String? = null
) : StoredDocumentDto, ICureDocumentDto<String>, EncryptableDto {
	override fun withIdRev(id: String?, rev: String) = if (id != null) this.copy(id = id, rev = rev) else this.copy(rev = rev)
	override fun withDeletionDate(deletionDate: Long?) = this.copy(deletionDate = deletionDate)
}
